### PR TITLE
fix(helm-requirements): remove the need for a Chart.yaml file

### LINKED ISSUE #7547: Remove the need of a Chart.yaml for helm-requirements manager
**What would you like Renovate to be able to do?**
To make the helm-requirements manager not need a `Chart.yaml` file

My personal context:
Mine is a bit of a special case. I have a custom format on top of helm, that just has a requirements.yaml file. I verified renovate works fine against it by commiting a dummy Chart.yaml. By looking at the code it seems that that file's contents don't make a difference, so I believe we could just remove its need from the codebase.

**Did you already have any implementation ideas?**
Yes I've implemented it here https://github.com/renovatebot/renovate/pull/7544

**Are there any workarounds or alternative ideas you've tried to avoid needing this feature?**
Yes, I could run some code before renovate that creates a dummy `Chart.yaml` file. Not entirely sure `renovate` supports that out of the box though.

**Is this a feature you'd be interested in implementing yourself?**
Yes, this is implemented in the PR linked.

### PR DESCRIPTION
**Pull request by eh-am**
Remove the need for the presence of a `Chart.yaml` file.

More context:
Mine is a bit of a special case. I have a custom format on top of helm, that just has a `requirements.yaml` file. I verified renovate works fine against it by commiting a dummy `Chart.yaml`. By looking at the code it seems that that file's contents don't make a difference, so I believe we could just remove its need from the codebase.

Closes #7547

### CODE REVIEW

**Comment by rarkins** [comment]
Hi, this needs discussion and agreement in an issue first.

**Comment by eh-am** [comment]
Makes sense, I assumed since it's a backwards compatible change we could skip the discussion process, but I was mistaken, my apologies. Here's the issue https://github.com/renovatebot/renovate/issues/7547

**Comment by rarkins** [comment]
Let's add some hardening around this part:

```
  deps = doc.dependencies.map((dep) => {
    const res: PackageDependency = {
      depName: dep.name,
      currentValue: dep.version,
    };
```

e.g. we should check that each dep contains both a `name` and `version` field.

**Comment by eh-am** [comment]
Hi @rarkins, I may have gone little overboard with the validations. Let me know what you think.

PS:
I haven't tried https://github.com/renovatebot/renovate/issues/7547#issuecomment-716198199 yet. That may invalidate the need for this PR.

**Comment by eh-am** [comment]
Hi @rarkins, any news on this?  Unfortunately, the regex approach suggested by https://github.com/renovatebot/renovate/issues/7547#issuecomment-716198199 didn't work as well we would like (it requires a certain order which we guarantee).

File: lib/manager/helm-requirements/extract.spec.ts [hunk 1]
```diff
@@ -1,147 +1,184 @@
...
+    it('parses simple requirements.yaml correctly', async () => {
       const content = `
       dependencies:
         - name: redis
           version: 0.9.0
           repository: https://kubernetes-charts.storage.googleapis.com/
         - name: postgresql
           version: 0.8.1
-          repository: file:///some/local/path/
+          repository: https://kubernetes-charts.storage.googleapis.com/
```
Comment (rarkins) [inline lib/manager/helm-requirements/extract.spec.ts:181]: Why was this test changed?

**Review by rarkins — CHANGES_REQUESTED** [review]
It's quite hard to evaluate this and be confident when there's so many changes. Is it possible to divide this into multiple PRs to merge one by one?

1. Add or change tests only with existing implementation passing.
2. Refactor existing implementation, leave tests unchanged.
3. Change implementation, add a few tests

The idea is to move as much of the changes as possible to refactoring PRs, where either (a) only tests change, or (b) only implementation changes. Then it's much easier to validate from inspection.

**Comment by eh-am** [comment]
Makes total sense. To be fair, my required changes are very simple: just remove a validation.

I think we can break into 2 steps:
1 - Remove the validation (looking up Chart.yaml) and the tests associated with that validation (latest commit)
2 - Harden the current implementation with https://github.com/renovatebot/renovate/pull/7544#issuecomment-717543737 and add associated tests, if necessary, more changes can be done, each on their own PR.

What do you think?

**Comment by rarkins** [comment]
Is there any reason not to add the hardening first?

**Comment by eh-am** [comment]
Ah, it's just that deleting code is easier than adding!

Here's the other PR
https://github.com/renovatebot/renovate/pull/8218

**Comment by eh-am** [comment]
It's just that deleting code is easier than adding!

I started the change in a branch. I have few questions though:

1 - What to do when there's semantic versioning prefix (~^). Should we ignore them or skip altogether?

2 - There can be multiple reasons to skip a dependency (missing name, version, and repository). Is there a way to return all of them? If not, should we pick an order? Which one?
Example: name first, then version, then repository.

3 - Should we test for a specific skipReason, or just for its presence?

3 - What's the stance on snapshot testing? I see them being used in multiple places, but it seems bit easy to overlook mistakes (I know I did).

**Comment by rarkins** [comment]
> It's just that deleting code is easier than adding!

Yes, but my concern is that if we merge this first then we are "unhardened" and lead or warnings/errors unncecessarily.

> I started the change in a branch. I have few questions though:
> 
> 1 - What to do when there's semantic versioning prefix (~^). Should we ignore them or skip altogether?

If we don't support ranges like that yet then we need to add a skipReason

> 2 - There can be multiple reasons to skip a dependency (missing name, version, and repository). Is there a way to return all of them? If not, should we pick an order? Which one?
> Example: name first, then version, then repository.

Yes, sounds fine

> 3 - Should we test for a specific skipReason, or just for its presence?
> As a user, I find `skipReason` to be useful information, so maybe we should test for it?

Sure, skipReason should be deterministic in such cases (based on the order above)

> 4 - What's the stance on snapshot testing? I see them being used in multiple places, but it seems bit easy to overlook mistakes (I know I did).

It's certainly a trade-off to use snapshots. If you have the time to write them with literals it's much better because it's harder for someone in future to break them and everyone misses it (especially as diffs an be hard to read).

**Comment by eh-am** [comment]
Got It. https://github.com/renovatebot/renovate/pull/8227

> If we don't support ranges like that yet then we need to add a skipReason

Upon some quick investigation, it seems it's set `SkipReason.UnsuportedValue` by the worker. 

https://github.com/renovatebot/renovate/blob/255632f00357930e4426012f3bad19b3c681095f/lib/workers/repository/process/lookup/index.ts#L376

**Comment by eh-am** [comment]
Hi @rarkins,
Now that #8227 has been merged, are there any updates to be done there?

**Comment by rarkins** [comment]
@eh-am if you resolve conflicts and mark as ready for review then we can take a look again

**Pull request merged**